Active Directory Login Integration

AnthonyCSheehy opened the pull request:
This addition based on the LDAP module adds basic Active Directory login integration support for logging in to AtoM.

Our Archivist wants users to use their existing credentials and auto-creation of accounts in AtoM for our ~740 pupils and ~300 staff. Importing is not possible as AD stores credentials in a non-reversible format as per policy. Only a handful of staff will be given additional permissions.

The Active Directory module differs from the LDAP module in only a handful of ways:

Allows for LDAPS using a URI (e.g. LDAPS://[:]) rather than a host and port. For this to work the CA Certificate of the issuing authority that issues the LDAP server's certificate must be entered in to the file specified in /etc/ldap/ldap.conf (or equivalent.)

Uses the UserPrincipalName (UPN) to identify the user to the AD server. This is the Windows 2003+ user login of @ that replaced the <username> a long time ago. This allows for identification of users throughout the directory, not just on a specified Organisational Unit within the directory.

Reduces the number of settings to 2 - the URI and the Search Base - which should be the root of the directory.

Uses the email address to store the UPN and identify the returning user. It pulls the name attribute from the directory to store as the user name. The additional benefit of this is that a user that is already created in AtoM will not have the account overridden or re-created (should the email address match the UPN.)

I created a separate class so that the LDAP module can continue to be developed and this class will not interfere with that Module.

AnthonyCSheehy committed:
Added adUser class

AnthonyCSheehy committed:
Added the adAction class

AnthonyCSheehy committed:
Added the ad Success file

AnthonyCSheehy committed:
Updated menuComponent class

AnthonyCSheehy commented:
Documentation:

Change the myUser to ADUser in /config/factories.yml. Clear the cache and recycle the services. A new section will then appear under settings.

1. DC or GC URI: This is the address of your Active Directory server in the format: [ldap|ldaps]://\<servername\>[:\<customport\>] The port number is only needed if it is non-standard. (Note: if using ldaps the CA certificate for the issuing CA should be added to the file specified in /etc/ldap/ldap.conf)

2. Search Base: This should be the root of your directory, for example DC=artefactual,DC=com

Users log in using their userPrincipalName (UPN) - the \<username\>@\<domain\> format. Users with their UPN already existing as an email address will not get overwritten.

AnthonyCSheehy committed:
Update adAction.class.php

mcantelon commented on apps/qubit/modules/settings/actions/adAction.class.php:
```diff
@@ -0,0 +1,99 @@
+ * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+class SettingsadAction extends DefaultEditAction
```
SettingsActiveDirectoryAction might be more explicit.

AnthonyCSheehy replied:
True, but then for consistency it might be best to change all the other instances of ad to activedirectory... It could be done. Do you think it would get confused with anything else?

mcantelon replied:
Yeah, other instances would have to be changed too. SettingsadAction isn't camel-case, so would be good to change it to something else. SettingsADAction could work too.

AnthonyCSheehy replied:
Hehe... I have learned a new phrase today... I have CamelCaseified the class names...

mcantelon commented on apps/qubit/modules/settings/actions/adAction.class.php:
```diff
@@ -0,0 +1,99 @@
+  {
+    switch ($name)
+    {
+
```
Extra blank line

AnthonyCSheehy replied:
Removed

mcantelon commented on apps/qubit/modules/settings/actions/adAction.class.php:
```diff
@@ -0,0 +1,99 @@
+        }
+        else
+        {
+
```
Extra blank line

AnthonyCSheehy replied:
Removed

mcantelon commented on apps/qubit/modules/settings/actions/adAction.class.php:
```diff
@@ -0,0 +1,99 @@
+  {
+    switch ($name = $field->getName())
+    {
+		case 'ldapHost':
```
Some excess indentation (maybe spaces/tab issue?)

AnthonyCSheehy replied:
Corrected...

mcantelon commented on apps/qubit/modules/settings/actions/menuComponent.class.php:
```diff
@@ -91,6 +91,16 @@ public function execute($request)
+        'action' => 'ad'
+      ));
+    }
+
```
Extra blank line

AnthonyCSheehy replied:
Removed

mcantelon commented on apps/qubit/modules/settings/templates/adSuccess.php:
```diff
@@ -0,0 +1,46 @@
+        <?php echo $form->ldapBaseDn
+          ->label(__('Base DN'))
+          ->renderRow() ?>
+
```
Extra blank line

AnthonyCSheehy replied:
Removed

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+    {
+      // Load user using username or, if one doesn't exist, create it
+      $criteria = new Criteria;
+      //$criteria->add(QubitUser::USERNAME, $username);
```
Probably best to remove these two commented-out lines.

AnthonyCSheehy replied:
Removed

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+    {
+      //In order to relocate the user when logging in again, the email must match the UPN
+      //which in some organisations will match the email address in any case.
+      //if (!empty($entries[0]['mail'])) $user->email = $entries[0]['mail'][0];
```
Commented-out code

AnthonyCSheehy replied:
Removed

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+    } else { $user->username = $username; }
+
+    $user->save();
+
```
Extra blank line

AnthonyCSheehy replied:
Removed

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    $host = QubitSetting::getByName('ldapHost');
+    //$port = QubitSetting::getByName('ldapPort');
```
Commented-out code

AnthonyCSheehy replied:
Removed

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+
+    if (null !== $host)
+    {
+		
```
Extra blank

AnthonyCSheehy replied:
Removed

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+    if (null !== $host)
+    {
+		
+			// If using an URI you only need to send the host URI, so the $port will be null
```
Extra indentation in this block of code

AnthonyCSheehy replied:
Corrected

mcantelon commented on lib/adUser.class.php:
```diff
@@ -0,0 +1,192 @@
+          
+      		$this->ldapConnection = $connection;
+      		return $connection;
+		//}
```
Commented out code

AnthonyCSheehy replied:
Removed

mcantelon reviewed (changes requested):
Other than a number of cosmetic changes, looks good!

AnthonyCSheehy committed:
Update adAction.class.php

Cleaned up indentation and removed unnecessary blank lines.

AnthonyCSheehy committed:
Update menuComponent.class.php

Removed extra blank line.

AnthonyCSheehy committed:
Update adSuccess.php

Removed extra blank line.

AnthonyCSheehy committed:
Update adUser.class.php

Removed additional blank lines and corrected tabulation. Removed commented-out code.

AnthonyCSheehy committed:
Update adAction.class.php

Class name camelcase-ified...

AnthonyCSheehy committed:
Update adUser.class.php

CamelCase-ified the Class Name

AnthonyCSheehy committed:
Update menuComponent.class.php

CamelCase-ified the AD Class name.

mcantelon commented:
Unfortunately, we're not going to be able to add this functionality at the present time as there are concerns about our ability to support and maintain it (as we have no Active Directory expertise). Our apologies for this, given you've taken the time to sign the contributor agreement and prepare the pull request. Thank you for submitting this... people who take the time to contribute are rare and we appreciate it.

fiver-watson commented:
Hi Anthony, 

Apologies for the mixed messages you have received. You've done great work here and been very responsive on the pull request (which is very appreciated!), but some of our managers feel that we shouldn't accept code when we are unable to thoroughly test, document, and maintain it through subsequent releases. This is out of my hands, unfortunately. 

I do think that there is community interest around this, and if you would like, I would be happy to discuss ways in which we could share your work with the community - perhaps adding a link to https://github.com/AnthonyCSheehy/atom and further information on our wiki for users would like to pull in your changes themselves? We could also update the various threads in the forum about ActiveDirectory with a link. Open to other thoughts as well. 

Regards, 

Dan Gillean
AtoM Program Manager
dan@artefactual.com

AnthonyCSheehy commented:
Hey Thanks. Okay - no problem. Happy to do that. I'll get started by making
sure that fork is updated.

On 22 January 2018 at 15:39, Dan Gillean <notifications@github.com> wrote:

> Hi Anthony,
>
> Apologies for the mixed messages you have received. You've done great work
> here and been very responsive on the pull request (which is very
> appreciated!), but some of our managers feel that we shouldn't accept code
> when we are unable to thoroughly test, document, and maintain it through
> subsequent releases. This is out of my hands, unfortunately.
>
> I do think that there is community interest around this, and if you would
> like, I would be happy to discuss ways in which we could share your work
> with the community - perhaps adding a link to https://github.com/
> AnthonyCSheehy/atom and further information on our wiki for users would
> like to pull in your changes themselves? We could also update the various
> threads in the forum about ActiveDirectory with a link. Open to other
> thoughts as well.
>
> Regards,
>
> Dan Gillean
> AtoM Program Manager
> dan@artefactual.com
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/artefactual/atom/pull/641#issuecomment-359461621>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AhgieYzeIXKt9rybg5WG0-CQmFHUOosiks5tNKumgaJpZM4RUeHG>
> .
>

rferris commented:
Just wanted to chime in that Anthony's changes are the only way I could get authentication working with our AD server (with an LDAP interface). It's a shame that they couldn't be included in the base Atom product, as it's something I will be looking to use for the foreseeable future. Thanks Anthony!
